src/mlpack/tests/bayesian_linear_regression_test.cpp

```diff
@@ -0,0 +1,185 @@
+BOOST_AUTO_TEST_CASE(TestCenterDataTrueScaleDataTrue)
+{
+  arma::mat matX;
+  arma::rowvec y;
+  size_t nDims = 5, nPoints = 100;
+  GenerateProblem(matX, y, nPoints, nDims, 0.5);
+
+  BayesianLinearRegression estimator(true, true);
+  estimator.Train(matX, y);
+
+  arma::colvec xMean = arma::mean(matX, 1);
+  arma::colvec xStd = arma::stddev(matX, 0, 1);
+  double yMean = arma::mean(y);
+
+  BOOST_REQUIRE_SMALL((double) abs(sum(estimator.DataOffset() - xMean)), 1e-6);
+  BOOST_REQUIRE_SMALL((double) abs(sum(estimator.DataScale() - xStd)), 1e-6);
+  BOOST_REQUIRE_CLOSE(estimator.ResponsesOffset(), yMean, 1e-6);
+}
+
+// Make sure a model with center ans scale option set is different than a model
+// without it set.
+BOOST_AUTO_TEST_CASE(OptionsMakeModelDifferent)
+{
+  arma::mat matX;
+  arma::rowvec y;
+  size_t nDims = 10, nPoints = 100;
+  GenerateProblem(matX, y, nPoints, nDims, 0.5);
+
+  BayesianLinearRegression blr(false, false), blrC(true, false),
+      blrCS(true, true);
+
+  blr.Train(matX, y);
+  blrC.Train(matX, y);
+  blrCS.Train(matX, y);
+
+  for (size_t i = 0; i < nDims; ++i)
+    BOOST_REQUIRE((blr.Omega()(i) != blrC.Omega()(i)) &&
+                  (blr.Omega()(i) != blrCS.Omega()(i)) &&
+                  (blrC.Omega()(i) != blrCS.Omega()(i)));
+}
+
+// Check that Train() does not fail with two colinear vectors.
+BOOST_AUTO_TEST_CASE(SingularMatix)
+{
+  arma::mat matX;
+  arma::rowvec y;
+
+  GenerateProblem(matX, y, 200, 10);
+  // Now the first and the second rows are indentical.
+  matX.row(1) = matX.row(0);
+
+  BayesianLinearRegression estimator;
+  estimator.Train(matX, y);
+}
+
+// Check that std are well computed/coherent. At least higher than the
+// estimated predictive variance.
+BOOST_AUTO_TEST_CASE(PredictiveUncertainties)
+{
+  arma::mat matX;
+  arma::rowvec y;
+
+  GenerateProblem(matX, y, 100, 10, 1);
+
+  BayesianLinearRegression estimator(true, true);
+  estimator.Train(matX, y);
+
+  arma::rowvec responses, std;
+  estimator.Predict(matX, responses, std);
+  const double estStd = sqrt(estimator.Variance());
+
+  for (size_t i = 0; i < matX.n_cols; i++)
+    BOOST_REQUIRE_GT(std[i], estStd);
+
+  // Check that the estimated variance is close to 1.
+  BOOST_REQUIRE_CLOSE(estStd, 1, 20);
+}
+
+// Check the solution is equal to the classical ridge.
+BOOST_AUTO_TEST_CASE(EqualtoRidge)
+{
+  arma::mat matX;
+  arma::rowvec y;
+
+  GenerateProblem(matX, y, 100, 10, 1);
+
+  BayesianLinearRegression blr(false, false);
+  blr.Train(matX, y);
+
+  LinearRegression ridge(matX,
+                         y,
+                         blr.Alpha() / blr.Beta(),
+                         false);
+
+  arma::rowvec blrPred, ridgePred;
+  blr.Predict(matX, blrPred);
+  ridge.Predict(matX, ridgePred);
+
+  // Check the predictions are close enough between ridge an or tested model.
+  for (size_t i = 0; i < y.size(); ++i)
+    BOOST_REQUIRE_CLOSE(blrPred[i], ridgePred[i], 1);
```

Comment:
Just once in my 1000 runs, I saw a failure on this test:

```
/home/ryan/src/mlpack/src/mlpack/tests/bayesian_linear_regression_test.cpp(183): fatal error: in "BayesianLinearRegressionTest/EqualtoRidge": difference{0.226143} between blrPred[i]{9.0707706257298071e-07} and ridgePred[i]{7.3978084147441336e-07} exceeds 1%
```

I'm not sure why this would have happened.  My best guess is that computing the direct inversion in `LinearRegression` is giving different results, probably because the condition number of the matrix is very large.  Could this happen, perhaps, when the covariance ends up being very close to low-rank from `GenerateProblem()`?  I am not sure exactly how that could happen.  Anyway, maybe it makes sense to perform a test and just try again, like this:

```
for (size_t trial = 0; trial < 3; ++trial)
{
  GenerateProblem(matX, y, 100, 10, 1);

  BayesianLinearRegression blr(false, false);
  blr.Train(matX, y);

  LinearRegression ridge(matX,
                         y,
                         blr.Alpha() / blr.Beta(),
                         false);

  arma::rowvec blrPred, ridgePred;
  blr.Predict(matX, blrPred);
  ridge.Predict(matX, ridgePred);

  // If the predictions seem far off, just try again.
  if (arma::norm(blrPred - ridgePred) > 1e-5)
    continue;

  // Check the predictions are close enough between ridge an or tested model.
  for (size_t i = 0; i < y.size(); ++i)
    BOOST_REQUIRE_CLOSE(blrPred[i], ridgePred[i], 1);

  // Exit once a test case has completed.
  break;
}
```

Let me know what you think; maybe that's not the best possible solution.

Reply:
I think it is good like that.